### ISSUE TITLE
Error message with error popup"failed to allocate memory for thumbnail export, please lower the threads used for export or buy more memory"

### ISSUE THREAD

**Issue by ydyote** [darktable-org/darktable#4043]
**Describe the bug**
I want to import a folder recursively in lighttable and if the number of images is too high it will crash with message "failed to allocate memory for thumbnail export, please lower the threads used for export or buy more memory".

I'm not even sure to which memory is this message referring to graphic card memory or standard RAM?

Also it spawns a small window, in which when I click OK it will close darktable.

If you don't click the button and just wait, the import will finish.

(maybe this is a gpu issue, because I think it happens when the processing load is too high)

**To Reproduce**
1. Have ~10K images in multiple folders under 1 folder named "images".
2. Go to import -> folder
3. Check import recursively.
4. Select the "images" folder and click open.
5. Wait for a second and it should spawn the error window.

**Expected behavior**
I would expect at least clear steps to fix this on my side, if it is specific to my installation and PC.

**Screenshots + Files**

![image](https://user-images.githubusercontent.com/54998537/71813154-44b35180-3079-11ea-9af7-56eb2534f74d.png)
![image](https://user-images.githubusercontent.com/54998537/71813164-4a109c00-3079-11ea-8b65-ebcbf4620e0c.png)
[darktable_bt_THRVD0.txt](https://github.com/darktable-org/darktable/files/4025737/darktable_bt_THRVD0.txt)

**Platform (please complete the following information):**
 - Darktable Version: 3.0.0
 - OS: Windows 10 Pro v1809 build 17763.914
 - OpenCL activated, very fast GPU profile

CPU:           AMD Phenom II X4 955 Black Edition (Deneb, RB-C3 (Ridgeback))
               3700 MHz (18.50x200.0) @ 3709 MHz (18.50x200.5)
Motherboard:   MSI 770-C45 (MS-7599)
Chipset:       AMD 770/870 (RX780) + SB750/SB710
Memory:        16384 MBytes @ 668 MHz, 9-9-9-24
               - 4096 MB PC10600 DDR3 SDRAM - Kingston 9905471-006.A02LF
               - 4096 MB PC10600 DDR3 SDRAM - Kingston 9905471-006.A02LF
               - 4096 MB PC10600 DDR3 SDRAM - Kingston 99U5474-028.A00LF
               - 4096 MB PC10600 DDR3 SDRAM - Kingston 99U5474-028.A00LF
Graphics:      MSI N750Ti OC (MS-V8A9)
               NVIDIA GeForce GTX 750 Ti, 2048 MB GDDR5 SDRAM

**Additional context**
 - This was also a problem in previous darktable version.
 - I removed every configuration file which I found before installing 3.0.0
 - I didn't find this bug nowhere so thats why I created it here.

PS: What is this site "https://redmine.darktable.org/"? and why cannot I register, if it is inactive please redirect users somehow directly here, because I just put "darktable bug" and went to that site because it was the first one which I thought is active.

Keep up the good work, darktable is really good, Thank you all.

**Comment by ydyote** [comment]
Turning off OpenCL didn't help.

**Comment by junkyardsparkle** [comment]
> I would expect at least clear steps to fix this on my side, if it is specific to my installation and PC.

Did you try lowering the number of background threads in settings, as suggested in the error message?

**Comment by ydyote** [comment]
ok, now I tried lowering the number of background threads which is not exactly what is written in the error. 

settings -> core options -> number of background threads

for 1, 2 - number of background threads, the import works and the memory consumption of darktable is under 300MB, with 2 threads the whole pc slows down, but that is maybe only problem with my setup, but it works.

for 3, 4, 8 - number of background threads the memory just grows over 9GB and then it crashes.

I attached also my cpu / gpu / memory core settings, maybe it helps. (I only changed the number of background threads and always restarted darktable)
![image](https://user-images.githubusercontent.com/54998537/71921423-26d51200-3189-11ea-9dfc-f9c6133d4e3c.png)

**Comment by jenshannoschwalm** [comment]
I really don't concider this a dt issue. Still a few remarks, your system is simply not meant to be used with more than maybe 2 dt threads, more threads will definitely slow down the whole system. Or the windows scheduler gets wild (you noticed and ends up just consuming memory).

You should never use more threads than physical cores basically. So what you see is not a dt issue but your system running wild on this workload.

Point 2, you setup dt to use opencl and very fast GPU, for that system it will work for most cases but for that graphics card it will probably end up slowing the dt performance.

Please close issue, btw there can be found more information about such issues in the dt manual...

**Comment by ydyote** [comment]
My CPU is AMD Phenom II X4 955 Black Edition, the X4 stands for 4 physical cores btw, so I would maybe think its some weird calculations for some buffer size which will be set to infinite instead of fixed value - which will cause the memory leak? Don't forget this is on a Windows 10, is there nobody else who can also try this on his machine if it is PC specific or general bug?

nevertheless, imo no matter what user sets there it should not crash the whole application - or it should be hidden in some kind of advanced settings tab which will have warning when you change values there.

**Comment by jenshannoschwalm** [comment]
Maybe related to #4056? @peterbud any idea?

**Comment by jenshannoschwalm** [comment]
Hi @ydyote i started to dig into this. My hypothesis atm is the memory management for dt may be mislead by certain configure parameters involved here.

I am still finding out how to solve that and need some more help from you.

Your configure parameters seem quite excessive for your system, could you check what happens after setting to the following suggestions.

1. "host memory limit for tiling" seems to be pretty high for your system. Probably 500-1000 should be right
2. "minimum amount of memory" might also be a bit high.
3. A thumbnail cache of 2GB takes away 1/8 of your whole memory, that might also be quite high.
4. Do you have "color manage cached thumbnails" switched on?
5. Do you have "don't use embedded preview JPEG but half-size raw" switched on? 

it would be really helpful if you could check!

**Comment by ydyote** [comment]
4. yes
5. no

![image](https://user-images.githubusercontent.com/54998537/72989919-7ead8300-3def-11ea-8973-c87e53be58e5.png)

Right now I could not reproduce the crash anymore, RAM usage by darktable is 6GB and doesn't go down so I'm not sure what happened. It also increases for every import I run but it is maybe like 100mb increments.
I will try to help again if it crashes for me when using 4 threads (I don't know if it will happen anytime soon, sorry).

**Comment by jenshannoschwalm** [comment]
@ydyote This issue has likely been solved, could you check again and either close this issue or report?

**Comment by hrehfeld** [comment]
I'm also suffering from constant (unusable state) memory allocation failures in various forms, and interestingly, I'm **also on an AMD Phenom II** (X6 1055t, 6 cores) [8GB RAM] with NVIDIA (1060) and similar settings (set as recommended in https://darktable.gitlab.io/doc/en/special_topics_chapter.html#darktable_and_memory_64bit ).

I'm using **todays git build** via AUR package: `darktable 3.1.0+456~gc5536e0b2`. Probably after the fix?

I tried increasing `ulimits` -- before it wasn't even starting up with all sort of memory alloc failures -- but it still crashes all the time:

```
$ ulimit -a
-t: cpu time (seconds)              unlimited
-f: file size (blocks)              unlimited
-d: data seg size (kbytes)          unlimited
-s: stack size (kbytes)             32768
-c: core file size (blocks)         unlimited
-m: resident set size (kbytes)      unlimited
-u: processes                       65000
-n: file descriptors                65536
-l: locked-in-memory size (kbytes)  1048576
-v: address space (kbytes)          8388608
-x: file locks                      unlimited
-i: pending signals                 31529
-q: bytes in POSIX msg queues       819200
-e: max nice                        0
-r: max rt priority                 0
-N 15:                              unlimited
```

### Also, startup time is horrible -- 5minute-ish. It hangs after printing:

```
$ darktable -d memory
[memory] at startup
[memory] max address space (vmpeak):      106380 kB
[memory] cur address space (vmsize):      106380 kB
[memory] max used memory   (vmhwm ):       23016 kB
[memory] cur used memory   (vmrss ):       23016 kB
[memory] after successful startup
[memory] max address space (vmpeak):     8334756 kB
[memory] cur address space (vmsize):     8334756 kB
[memory] max used memory   (vmhwm ):      327044 kB
[memory] cur used memory   (vmrss ):      307036 kB
```
This also happens when I delete memory options from `darktablerc` and with worker threads set to 1.

### The memory allocation failures come in various forms:

a)
I seem to be getting these with number of worker threads different from 8?!
```
$ darktable -d memory
[memory] at startup
[memory] max address space (vmpeak):      106380 kB
[memory] cur address space (vmsize):      106380 kB
[memory] max used memory   (vmhwm ):       22724 kB
[memory] cur used memory   (vmrss ):       22724 kB

libgomp: Thread creation failed: Resource temporarily unavailable
```
This also happens when I delete memory options from `darktablerc` and with worker threads set to 1.

b)
```
$ darktable -d opencl
0.201585 [opencl_init] opencl related configuration options:
0.201612 [opencl_init] 
0.201618 [opencl_init] opencl: 1
0.201623 [opencl_init] opencl_scheduling_profile: 'very fast GPU'
0.201628 [opencl_init] opencl_library: ''
0.201634 [opencl_init] opencl_memory_requirement: 768
0.201645 [opencl_init] opencl_memory_headroom: 400
0.201650 [opencl_init] opencl_device_priority: '*/!0,*/*/*/!0,*'
0.201656 [opencl_init] opencl_mandatory_timeout: 200
0.201665 [opencl_init] opencl_size_roundup: 16
0.201681 [opencl_init] opencl_async_pixelpipe: 0
0.201686 [opencl_init] opencl_synch_cache: active module
0.201691 [opencl_init] opencl_number_event_handles: 25
0.201704 [opencl_init] opencl_micro_nap: 1000
0.201715 [opencl_init] opencl_use_pinned_memory: 0
0.201724 [opencl_init] opencl_use_cpu_devices: 0
0.201734 [opencl_init] opencl_avoid_atomics: 0
0.201742 [opencl_init] 
0.202259 [opencl_init] found opencl runtime library 'libOpenCL'
0.202317 [opencl_init] opencl library 'libOpenCL' found on your system and loaded
0.241185 [opencl_init] found 1 platform
0.241214 [opencl_init] found 1 device
0.241522 [opencl_init] device 0 `GeForce GTX 1060 6GB' has sm_20 support.
0.241793 [opencl_init] device 0 `GeForce GTX 1060 6GB' supports image sizes of 16384 x 32768
0.241801 [opencl_init] device 0 `GeForce GTX 1060 6GB' allows GPU memory allocations of up to 1519MB
[opencl_init] device 0: GeForce GTX 1060 6GB 
     GLOBAL_MEM_SIZE:          6078MB
     MAX_WORK_GROUP_SIZE:      1024
     MAX_WORK_ITEM_DIMENSIONS: 3
     MAX_WORK_ITEM_SIZES:      [ 1024 1024 64 ]
     DRIVER_VERSION:           440.48.02
     DEVICE_VERSION:           OpenCL 1.2 CUDA
0.607365 [opencl_init] could not create context for device 0: -6
0.607397 [opencl_init] no suitable devices found.
0.607400 [opencl_init] FINALLY: opencl is NOT AVAILABLE on this system.
0.607403 [opencl_init] initial status of opencl enabled flag is OFF.
```
on my system: 
```
$ ag '\-6' /usr/include/gegl-0.4/opencl/cl.h
101:#define CL_OUT_OF_HOST_MEMORY                       -6
```

c)
```
[memory] max used memory   (vmhwm ):      820068 kB
[memory] cur used memory   (vmrss ):      332096 kB
[memory] before pixelpipe process
[memory] max address space (vmpeak):     8388408 kB
[memory] cur address space (vmsize):     7460760 kB
[memory] max used memory   (vmhwm ):      820068 kB
[memory] cur used memory   (vmrss ):      308412 kB
[memory] before pixelpipe process
[memory] max address space (vmpeak):     8388408 kB
[memory] cur address space (vmsize):     7451984 kB
[memory] max used memory   (vmhwm ):      820068 kB
[memory] cur used memory   (vmrss ):      327756 kB
[memory] before pixelpipe process
[memory] max address space (vmpeak):     8388408 kB
[memory] cur address space (vmsize):     7500660 kB
[memory] max used memory   (vmhwm ):      820068 kB
[memory] cur used memory   (vmrss ):      384316 kB
[memory] before pixelpipe process
[memory] max address space (vmpeak):     8388408 kB
[memory] cur address space (vmsize):     7454392 kB
[memory] max used memory   (vmhwm ):      820068 kB
[memory] cur used memory   (vmrss ):      344568 kB
[memory] before pixelpipe process
[memory] max address space (vmpeak):     8388408 kB
[memory] cur address space (vmsize):     7455820 kB
[memory] max used memory   (vmhwm ):      820068 kB
[memory] cur used memory   (vmrss ):      347992 kB
[memory] before pixelpipe process
[memory] max address space (vmpeak):     8388408 kB
[memory] cur address space (vmsize):     8303904 kB
[memory] max used memory   (vmhwm ):      820068 kB
[memory] cur used memory   (vmrss ):      586136 kB
[memory] before pixelpipe process
[memory] max address space (vmpeak):     8388408 kB
[memory] cur address space (vmsize):     7458032 kB
[memory] max used memory   (vmhwm ):      820068 kB
[memory] cur used memory   (vmrss ):      346016 kB
[memory] before pixelpipe process
[memory] max address space (vmpeak):     8388408 kB
[memory] cur address space (vmsize):     7531484 kB
[memory] max used memory   (vmhwm ):      820068 kB
[memory] cur used memory   (vmrss ):      363656 kB                                                                                                                                            
[memory] before pixelpipe process                                                                                                                                                              
[memory] max address space (vmpeak):     8388408 kB                                                                                                                                            
[memory] cur address space (vmsize):     8364028 kB                                                                                                                                            
[memory] max used memory   (vmhwm ):      820068 kB                                                                                                                                            
[memory] cur used memory   (vmrss ):      712816 kB                                                                                                                                            
[memory] before pixelpipe process                                                                                                                                                              
[memory] max address space (vmpeak):     8388408 kB                                                                                                                                            
[memory] cur address space (vmsize):     8284344 kB                                                                                                                                            
[memory] max used memory   (vmhwm ):      895940 kB                                                                                                                                            
[memory] cur used memory   (vmrss ):      591712 kB                                                                                                                                            
[memory] before pixelpipe process                                                                                                                                                              
[memory] max address space (vmpeak):     8388408 kB                                                                                                                                            
[memory] cur address space (vmsize):     7585428 kB                                                                                                                                            
[memory] max used memory   (vmhwm ):      895940 kB                                                                                                                                            
[memory] cur used memory   (vmrss ):      515968 kB                                                                                                                                            
[memory] before pixelpipe process                                                                                                                                                              
[memory] max address space (vmpeak):     8388408 kB                                                                                                                                            
[memory] cur address space (vmsize):     8362480 kB                                                                                                                                            
[memory] max used memory   (vmhwm ):      895940 kB                                                                                                                                            
[memory] cur used memory   (vmrss ):      681896 kB                                                                                                                                            
[memory] before pixelpipe process                                                                                                                                                              
[memory] max address space (vmpeak):     8388408 kB                                                                                                                                            
[memory] cur address space (vmsize):     8281980 kB                                                                                                                                            
[memory] max used memory   (vmhwm ):      895940 kB                                                                                                                                            
[memory] cur used memory   (vmrss ):      618888 kB                                                                                                                                            
zsh: segmentation fault (core dumped)  darktable -d memory             
```

etc.

During startup:
```
$ free                                     
              total        used        free      shared  buff/cache   available
Mem:        8154228     3192892     1248940      266496     3712396     4395160
Swap:       8388604       93696     8294908

```

This happens with archlinux stable binary (3.x) as well as todays git:

```
$ darktable --version
this is darktable 3.1.0+456~gc5536e0b2
copyright (c) 2009-2020 johannes hanika
darktable-dev@lists.darktable.org

compile options:
  bit depth is 64 bit
  normal build
  SSE2 optimized codepath enabled
  OpenMP support enabled
  OpenCL support enabled
  Lua support enabled, API version 5.0.2
  Colord support enabled
  gPhoto2 support enabled
  GraphicsMagick support enabled
  OpenEXR support enabled
```

No other apps seem to be having the same problems.

Help!

**Comment by jenshannoschwalm** [comment]
I think this is unrelated to the "fixing pr", probably you should open another issue.

Also there are a few messages that worry me about your systems configuration.

> libgomp: Thread creation failed: Resource temporarily unavailable

> 0.607365 [opencl_init] could not create context for device 0: -6
0.607397 [opencl_init] no suitable devices found.

**Comment by jenshannoschwalm** [comment]
@TurboGit , this has been fixed via #4210

**Comment by TurboGit** [comment]
Ok, closing. Thanks.
